.github/workflows/ci.yml

```diff
@@ -0,0 +1,127 @@
+name: ci
+
+on: [push, pull_request]
+
+env:
+  NODE_VERSION: 10
+  SOLR_VERSION: 6.6.5
+
+jobs:
+
+  build-static:
+    name: Build static files
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v2
+    - name: Setup node
+      uses: actions/setup-node@v2-beta
+      with:
+        node-version: ${{ env.NODE_VERSION }}
+    - name: Cache node modules
+      uses: actions/cache@v2
+      with:
+        path: ~/.npm
+        key: npm-${{ hashFiles('package-lock.json') }}
+        restore-keys: |
+          npm-${{ hashFiles('package-lock.json') }}
+          npm-
+    - name: Install dependencies
+      run: npm ci
+    - name: Run webpack build
+      run: npm run build:prod
+    - name: Upload built files
+      uses: actions/upload-artifact@v2
+      with:
+        name: bundles
+        path: bundles/
+    - name: Upload stats file
+      uses: actions/upload-artifact@v2
+      with:
+        name: webpack-stats.json
+        path: webpack-stats.json
+
+  ts-unit:
+    name: TypeScript unit tests
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v2
+    - name: Setup node
+      uses: actions/setup-node@v2-beta
+      with:
+        node-version: ${{ env.NODE_VERSION }}
+    - name: Cache node modules
+      uses: actions/cache@v2
+      with:
+        path: ~/.npm
+        key: npm-${{ hashFiles('package-lock.json') }}
+        restore-keys: |
+          npm-${{ hashFiles('package-lock.json') }}
+          npm-
+    - name: Install dependencies
+      run: npm ci
+    - name: Run Jest
+      run: npx jest --collectCoverage
+    - name: Upload test coverage to Codecov
+      uses: codecov/codecov-action@v1
+      with:
+        flags: typescript
+
+  py-unit:
+    name: Python unit tests
+    runs-on: ubuntu-latest
+    needs: build-static
+    strategy:
+      matrix:
+        python: [3.5, 3.6, 3.7]
+    services:
+      db:
+        image: mariadb
+        env:
+          MYSQL_DATABASE: sandco
+          MYSQL_ROOT_PASSWORD: sandco
+        ports:
+        - 3306:3306
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v2
+    - name: Setup Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python }}
+    - name: Cache pip
```

Comment:
What does caching pip get us?

Reply:
just speed. most of our dependencies shouldn't change too much between runs, so it saves hitting pypi and redownloading everything. you can see a little further on in this config that the caching strategy uses fallbacks; it will try to pull the versions corresponding to a hash of `requirements.txt` first (i.e. if there was no change at all) before it tries some other things and then eventually downloads when necessary.